orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -1129,6 +1129,86 @@ Example 4:
         }
       });
       
+    ONNX_CONTRIB_OPERATOR_SCHEMA(SplitTraining)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
+      .SetDoc("SplitTraining")
+      .Attr("axis",
+            "Which axis to split on. "
+            "A negative value means counting dimensions from the back. Accepted range is [-rank, rank-1] "
+            "where r = rank(input).",
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+      .AllowUncheckedAttributes()
+      .Input(0, "input", "The tensor to split", "T")
+      .Input(1, "split", "length of each output", "tensor(int64)")
+      .Output(0,
+              "outputs",
+              "One or more outputs forming list of tensors after splitting",
+              "T",
+              OpSchema::Variadic)
+      .TypeConstraint(
+          "T",
+          OpSchema::all_tensor_types(),
+          "Constrain input and output types to all tensor types.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        for (int i = 0; i < static_cast<int>(ctx.getNumOutputs()); ++i) {
+          propagateElemTypeFromInputToOutput(ctx, 0, i);
+        }
+        if (!hasNInputShapes(ctx, 1)) {
+          return;
+        }
+
+        // skip if split is not an initializer
+        auto split_proto = ctx.getInputData(1);
+        if (split_proto == nullptr) {
+          return;
+        }
+        std::vector<int64_t> split = ParseData<int64_t>(split_proto);
+        
+        if (!ctx.getInputType(0)->tensor_type().has_shape()) {
+          return;
+        }
+        const auto& shape = ctx.getInputType(0)->tensor_type().shape();
+        int rank = shape.dim_size();
+        int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
+        if (axis < -rank || axis >= rank) {
+          fail_type_inference(
+              "Invalid value of attribute 'axis'. Rank=",
+              rank,
+              " Value=",
+              axis);
+        }
+        if (axis < 0) {
+          axis += rank;
+        }
+        const auto& splitDim = shape.dim(axis);
+        if (!splitDim.has_dim_value()) {
+          return;
+        }
+        int splitDimValue = static_cast<int>(splitDim.dim_value());
+        int chunkSize =
```

Comment:
this implies that we are splitting the input equally over num_outputs. We shouldnt be modifying the 'split' vector in the next few lines, but use the value as is to set the output shape

Reply:
in fact it will not break the logic, if split = {1,3}. here we 'll get {1,3,2,2}, the last 2 won't be used below.
but I am adding the check on split to skip the logic if it's not empty

---
In reply to: [459790937](https://github.com/microsoft/onnxruntime/pull/4597#discussion_r459790937) [](ancestors = 459790937)